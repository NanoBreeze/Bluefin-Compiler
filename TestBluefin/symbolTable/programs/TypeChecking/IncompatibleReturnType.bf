/*
The type of some of the returned variables aren't compatible with the return type of the function
*/

void foo() {
	return 1; // can't return value
}

void zoo() {
	return; // no problem
}

int bar() {
	return; // missing value here. This is like returning 'void', which doesn't match with int
}

int car() { // missing return statement in direct function

	if (true) {
		return 1;	
	}
}

struct A {
	int one() {
		return true; // true doesn't convert to int
	}

	float two() {
		return 4; // type promotion, ok	
	}

	float three() {
		return 4*true+2; // not okay, the expr corresponding to '4*true+2' is not valid	
	}
};

struct B {};

A returnA() {
	B b;
	return b; // Since B doesn't inherit from A, this is invalid
}
