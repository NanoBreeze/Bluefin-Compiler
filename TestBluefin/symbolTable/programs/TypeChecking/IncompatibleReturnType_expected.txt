Return type invalid. Expected: 'void', Found: 'int'
Return type invalid. Expected: 'int', Found: 'void'
Return statement missing. Expected: 'int'
Return type invalid. Expected: 'int', Found: 'bool'
Operator '*' cannot be applied to 'int' and 'bool'
Return type invalid. Expected: 'A', Found: 'B'
